content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,326 @@
+```
+
+## Defining an Inferred Type
+
+The `typeof` function can be used in a type definition to define an inferred type
+
+```lua
+type Car = typeof {
+	Speed = 0,
+	Wheels = 4
+} --> Car: {Speed: number, Wheels: number}
+```
+
+One interesting use is to define a metatable type using `setmetatable` inside `typeof`
+
+```lua
+type Vector = typeof(setmetatable({}::{
+	x: number,
+	y: number
+}, {}::{
+	__add: (Vector, Vector|number) -> Vector
+}))
+
+-- Vector + Vector would return a Vector type
+```
+
+## Generics
+Generics allow you to define at a most basic level, substitution types. Lets assume a basic object that creates a State keyword
+
+```lua
+local State = {
+	Key = "TimesClickec",
+	Value = 0
+}
+```
+
+The type for this state object without generics would be
+```lua
+type State = {
+	Key: string,
+	Value: number
+}
+```
+
+However, you may want to `Value` to be a type based on the incoming value, this can be done with generics. Generic types use the following basic shape:
+
+```lua
+type GenericType<T> = T
+```
+
+The `<T>` denotes a type that can be set to anything. The best way to visualise this is as a substitution type. 
+
+```lua
+type List<T> = {T}
+
+local Names: List<string> = {"Bob", "Dan", "Mary"}
+local Fibbonacci: List<number> = {1, 1, 2, 3, 5, 8, 13}
+```
+
+Generics can also have multiple substitutions inside the brackets
+
+```lua
+type Map<K, V> = {[K]: V}
+```
+
+Using the state object above, it can be reworked to use a generic type as such
+
+```lua
+type State<T> = {
+	Key: string,
+	Value: T
+}
+```
+
+### Function Generics
+
+Functions can also use generics, however, they behave a little differently when inference is introduced into it
+
+Using the State example again, it can be designed to infer the value of T from the function's incoming arguments
+
+```lua
+local function State<T>(key: string, value: T): State<T>
+	return {
+		Key = key,
+		Value = value
+	}
+end
+
+local Activated = State("Activated", false) -- State<boolean>
+local TimesClicked = State("TimesClicked", 0) -- State<number>
+```
+
+### Type Packs
+
+Type packs are an extension to variadics that allow the definition of generic variadics inside functions.
+
+```lua
+type Event<A...> = {
+	Fire: (Event<A...>, A...)
+	Connect: (Event<A...>, func: (A...) -> ())
+}
+
+-- the event can be defined with as many types as you want
+local ChangedSignal: Event<string, any>
+
+ChangedSignal:Connect(function(changedName, changedValue)
+	-- changedName will be defined as a string
+	-- changedValue will be defined as any
+end)
+
+ChangedSignal:Fire("Size", Vector2.new(100, 50)) --> ok
+ChangedSignal:Fire(20, 100) --> not ok, number cannot be converted to string
+```
+
+Type packs are defined in functions using `...` on the right-hand side of a generic
+
+```lua
+local function Signal<A...>(f: (A...) -> (), ...: A...)
+```
+
+## Type Exports
+
+Types can be exported through `Class.ModuleScript` using the `export` keyword infront of a type definition

```

Comment:
```suggestion
Types can be exported through `Class.ModuleScript|ModuleScripts` using the `export` keyword infront of a type definition.
```

In the following code blocks, you may want to use titles and/or text to indicate that they belong to two different ModuleScripts rather than have the reader assume that from there being two separate code blocks. This will also convey the information better when localised so assumptions don't have to be made.

Reply:
I wonder how I could do this actually, I'm not really sure about this specific one

Reply:
You can assign titles to code blocks with the title parameter: `lua title="ModuleScript 1"`

Check out [scripts.md (raw)](https://raw.githubusercontent.com/Roblox/creator-docs/main/content/en-us/scripting/scripts.md) and the corresponding [scripts article on the Docs Site](https://create.roblox.com/docs/scripting/scripts) for an example of how it renders. "Example Module Script" is what you'll want to find.